src/middlewared/middlewared/etc_files/smb_configure.py

```diff
@@ -0,0 +1,352 @@
+import logging
+import os
+import time
+
+from middlewared.utils import run
+
+logger = logging.getLogger(__name__)
+
+NETCMD = "/usr/local/bin/net"
+PDBCMD = "/usr/local/bin/pdbedit"
+SMBPASSWDCMD = "/usr/local/bin/smbpasswd"
+TPCMD = "/usr/local/bin/testparm"
+
+TMP_PRIVATEDIR = "/root/private"
+TMP_SMBPASSWD = TMP_PRIVATEDIR + "/tmp_smbpasswd"
+
+"""
+    Preparations for starting samba. Needs to happen after smb4.conf is generated.
+    Goal is to:
+    1) Ensure that all required paths exist and are available.
+    2) Maintain local or domain SID
+    3) Validate and regenerate passdb.tdb if needed. Since entries are automatically
+       added as users are added in the UI, the primary situations where this is needed
+       are during HA failover and during boot.
+    4) Validate and regenerate local user group mapping.
+"""
+async def get_smb4conf_parm(parm):
+    tp = await run([TPCMD, "-s", f"--parameter-name={parm}"], check=False)
+    if tp.returncode != 0:
+        logger.debug(f'Command {TPCMD} for parameter {parm} failed with error: {tp.stderr.decode()}')
+        return False
+    parm_value = tp.stdout.decode().rstrip()
+
+    return parm_value
+
+async def get_config(middleware):
+    conf = {}
+    conf['systemdataset'] = await middleware.call('systemdataset.config')
+    # If the system dataset hasn't finished importing yet, exit quickly so that we don't perform unnecessary db operations.
+    if conf['systemdataset']['path'] is None:
+        return conf
+
+    conf['cifs'] = await middleware.call('smb.config')
+    conf['smb_users'] = await middleware.call('datastore.query', 'account.bsdusers', [
+        ['OR', [
+            ('bsdusr_smbhash', '~', r'^.+:.+:[X]{32}:.+$'),
+            ('bsdusr_smbhash', '~', r'^.+:.+:[A-F0-9]{32}:.+$'),
+        ]]
+    ])
+    dc = await middleware.call('datastore.query', 'services.services',
+                               [('srv_service', '=', 'domaincontroller')],
+                               {'get': True})
+    logger.debug(f'{dc}')
```

Comment:
Is this debug really necessary? If so perhaps it should be a trace log level? Dunno if logger.trace is available here.

Reply:
Will remove. Good catch. That's an artifact from testing.